File: apc-mcp/node/src/launcher/exampleWireup.ts
import { EventBus, TOPIC } from "../runtime/eventBus";
import { State, nextState } from "../runtime/stateMachine";
import { enforceToolPolicy } from "./middleware/criticalTool";

const bus = new EventBus();
let state = State.OK;

bus.on(TOPIC.CRC_SIGNAL, (msg:any) => {
  if (msg.kind === "drift" && msg.value > 0.10) {
    const to = nextState(state, "soft");
    bus.emit(TOPIC.NUDGE, { severity:"soft", message:"Focus drift detected", scope: msg.layer });
    bus.emit(TOPIC.STATE, { from: state, to, reason: "drift_soft" });
    state = to;
  }
});

// Example tool invocation interception
function runTool(call: {tool:any,args:any}) {
  const res = enforceToolPolicy(call as any, {
    allow: ["read","search","write","edit","bash","test_run"],
    maxTestTimeoutSec: 480
  });
  if (!res.ok) throw new Error(`PolicyDenied:${(res as any).reason}`);
  if (res.critical) {
    bus.emit(TOPIC.ENFORCE, { tool: call.tool, args: call.args, critical: true });
  }
  // ...execute tool here...
}

File: agentflow/apc_runtime/example_wireup.py
from agentflow.apc_runtime.event_bus import EventBus, TOPIC_CRC_SIGNAL, TOPIC_NUDGE, TOPIC_STATE
from agentflow.apc_runtime.state_machine import State, next_state
from agentflow.apc_runtime.lru import LRUCache

bus = EventBus()
state = State.OK
lru = LRUCache(capacity=512)

def on_crc_signal(msg):
    global state
    # msg: {layer, kind:'drift|fidelity|evidence', value:float, context_ref:str}
    if msg["kind"] == "drift" and msg["value"] > 0.10:
        s2 = next_state(state, "soft")
        bus.emit(TOPIC_NUDGE, {"severity":"soft","message":"Focus drift detected","scope":msg["layer"]})
        bus.emit(TOPIC_STATE, {"from":state, "to":s2, "reason":"drift_soft"})
        state = s2

bus.on(TOPIC_CRC_SIGNAL, on_crc_signal)
